File: PhaseJump/Xcode/PhaseJump/PhaseJumpTests/G Tests/Model/Time/TimerTests.cpp
#include "gtest/gtest.h"
#include "Timer.h"

#include <memory>

using namespace PJ;
using namespace std;

namespace TimerTests {
}

using namespace TimerTests;

TEST(Timer, RunOnce) {
    auto timer = Timer({
        .duration = 1.0f,
        .runType = RunType::Once
    });
    timer.OnUpdate(TimeSlice(.3f));
    EXPECT_EQ(.3f, timer.State());
    EXPECT_EQ(.3f, timer.Progress());
    EXPECT_FALSE(timer.IsFinished());

    timer.OnUpdate(TimeSlice(.7f));
    EXPECT_EQ(1.0f, timer.State());
    EXPECT_EQ(1.0f, timer.Progress());
    EXPECT_TRUE(timer.IsFinished());

    timer.OnUpdate(TimeSlice(.7f));
    EXPECT_EQ(1.0f, timer.State());
    EXPECT_EQ(1.0f, timer.Progress());
    EXPECT_TRUE(timer.IsFinished());

    timer.SetProgress(.5f);
    EXPECT_FALSE(timer.IsFinished());
}

TEST(Timer, Repeat) {
    int finishedCount = 0;
    
    auto timer = Timer({
        .duration = 1.0f,
        .runType = RunType::Repeat,
        .onFinishFunc = [&](auto timer) {
            finishedCount++;
        }
    });
    timer.OnUpdate(TimeSlice(.3f));
    EXPECT_EQ(0.3f, timer.State());
    EXPECT_EQ(0.3f, timer.Progress());

    timer.OnUpdate(TimeSlice(.7f));
    EXPECT_EQ(0, timer.State());
    EXPECT_EQ(0, timer.Progress());
    EXPECT_EQ(1, finishedCount);
    EXPECT_FALSE(timer.IsFinished());

    timer.OnUpdate(TimeSlice(.7f));
    EXPECT_EQ(0.7f, timer.State());
    EXPECT_EQ(0.7f, timer.Progress());
    EXPECT_FALSE(timer.IsFinished());

    timer.OnUpdate(TimeSlice(.3f));
    EXPECT_EQ(0, timer.State());
    EXPECT_EQ(0, timer.Progress());
    EXPECT_EQ(2, finishedCount);
    EXPECT_FALSE(timer.IsFinished());
}

TEST(Timer, Stop) {
    auto timer = Timer(1.0f, RunType::Once);
    timer.OnUpdate(TimeSlice(.3f));
    EXPECT_EQ(.3f, timer.State());
    EXPECT_EQ(.3f, timer.Progress());
    EXPECT_FALSE(timer.IsFinished());

    timer.Stop();
    EXPECT_TRUE(timer.IsFinished());
}

File: Experimental/Swift/PhaseJumpPro/Tests/PhaseJumpProTests/TimerTests.swift
import Foundation
import Testing
@testable import PhaseJumpPro

struct TimerTests {
    class Counter {
        var finishCount = 0
        var resetCount = 0
    }
    
    @Test
    func runOnce() async throws {
        let timer = Timer(1.0, runType: .once)
        timer.onUpdate(time: .init(delta: 0.3))
        #expect(0.3 == timer.state)
        #expect(0.3 == timer.progress)
        #expect(false == timer.isFinished)

        timer.onUpdate(time: .init(delta: 0.7))
        #expect(1.0 == timer.state)
        #expect(1.0 == timer.progress)
        #expect(timer.isFinished)

        timer.onUpdate(time: .init(delta: 0.7))
        #expect(1.0 == timer.state)
        #expect(1.0 == timer.progress)
        #expect(timer.isFinished)

        timer.progress = 0.5
        #expect(false == timer.isFinished)
    }

    @Test
    func testRepeat() async throws {
        let counter = Counter()
        
        let timer = Timer(1.0, runType: .repeat)
        timer.onFinishFunc = { _ in
            counter.finishCount += 1
        }
        timer.onUpdate(time: .init(delta: 0.3))
        #expect(0.3 == timer.state)
        #expect(0.3 == timer.progress)

        timer.onUpdate(time: .init(delta: 0.7))
        #expect(0 == timer.state)
        #expect(0 == timer.progress)
        #expect(1 == counter.finishCount)
        #expect(false == timer.isFinished)

        timer.onUpdate(time: .init(delta: 0.7))
        #expect(0.7 == timer.state)
        #expect(0.7 == timer.progress)
        #expect(false == timer.isFinished)

        timer.onUpdate(time: .init(delta: 0.3))
        #expect(0 == timer.state)
        #expect(0 == timer.progress)
        #expect(2 == counter.finishCount)
        #expect(false == timer.isFinished)
    }

    @Test
    func stop() {
        let timer = Timer(1.0, runType: .once)
        timer.onUpdate(time: .init(delta: 0.3))
        #expect(0.3 == timer.state)
        #expect(0.3 == timer.progress)
        #expect(false == timer.isFinished)

        timer.stop()
        #expect(timer.isFinished)
    }
}
